### PR TITLE
Add dump support

### PR DESCRIPTION
**Pull request by ppamorim**
Add dumps route support, unit and integration tests included.

### CODE REVIEW

**Comment by ppamorim** [comment]
bors try

File: Sources/MeiliSearch/Client.swift [hunk 1]
```diff
@@ -777,7 +779,40 @@ public struct MeiliSearch {
...
     }
+
+    /**
+     Triggers a dump creation process. Once the process is complete, a dump is created in the dumps folder.
+     If the dumps folder does not exist yet, it will be created.
+
+     - parameter completion: The completion closure used to notify when the server
+     completes the dump request, it returns a `Dump` object that contains `UID`
+     value that can be used later to check the status of the Dump.
+     If the request was sucessful or `Error` if a failure occured.
```
Comment (curquiza) [inline Sources/MeiliSearch/Client.swift:793]:
```suggestion
     If the request was successful or `Error` if a failure occurred.
```

File: Sources/MeiliSearch/Client.swift [hunk 2]
```diff
@@ -777,7 +779,40 @@ public struct MeiliSearch {
...
         self.system.version(completion)
     }
+
+    /**
+     Triggers a dump creation process. Once the process is complete, a dump is created in the dumps folder.
+     If the dumps folder does not exist yet, it will be created.
+
+     - parameter completion: The completion closure used to notify when the server
+     completes the dump request, it returns a `Dump` object that contains `UID`
+     value that can be used later to check the status of the Dump.
```
Comment (curquiza) [inline Sources/MeiliSearch/Client.swift:792]:
```suggestion
     value that can be used later to check the status of the dump.
```

File: Sources/MeiliSearch/Client.swift [hunk 3]
```diff
@@ -777,7 +779,40 @@ public struct MeiliSearch {
...
+    public func createDump(_ completion: @escaping (Result<Dump, Swift.Error>) -> Void) {
+        self.dumps.create(completion)
+    }
+
+    /**
+     Get the status of a dump creation process using the uid returned after calling the dump creation route.
+     The returned status could be:
+
+     `Dump.Status.processing`: Dump creation is in progress.
+     `Dump.Status.dumpProcessFailed`: An error occured during dump process, and the task was aborted.
```
Comment (curquiza) [inline Sources/MeiliSearch/Client.swift:804]:
```suggestion
     `Dump.Status.dumpProcessFailed`: An error occurred during the dump process, and the task was aborted.
```

File: Sources/MeiliSearch/Client.swift [hunk 4]
```diff
@@ -777,7 +779,40 @@ public struct MeiliSearch {
...
+     `Dump.Status.done`: Dump creation is finished and was successful.
+
+     - parameter completion: The completion closure used to notify when the server
+     completes the dump request, it returns a `Dump` object that contains `UID`
+     value that can be used later to check the status of the Dump.
+     If the request was sucessful or `Error` if a failure occured.
```
Comment (curquiza) [inline Sources/MeiliSearch/Client.swift:810]:
```suggestion
     If the request was successful or `Error` if a failure occurred.
```

File: Sources/MeiliSearch/Dumps.swift [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import Foundation
+
+/**
+ Settings is a list of all the customization possible for an index.
```
Comment (curquiza) [inline Sources/MeiliSearch/Dumps.swift:4]: Not the right info 😉

File: Tests/MeiliSearchIntegrationTests/DumpsTests.swift [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    func testCreateDump() {
+
+        let expectation = XCTestExpectation(description: "Request dump creation")
+
+        self.client.createDump { result in
+            switch result {
+            case .success(let dump):
+              XCTAssertTrue(!dump.UID.isEmpty)
+            case .failure:
+                XCTFail("Failed to request dump creation")
```
Comment (curquiza) [inline Tests/MeiliSearchIntegrationTests/DumpsTests.swift:29]: Weird indentation

File: Tests/MeiliSearchIntegrationTests/DumpsTests.swift [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        self.client.createDump { result in
+            switch result {
+            case .success(let createDump):
+
+                self.client.getDumpStatus(UID: createDump.UID) { result in
+                    switch result {
+                    case .success(let dumpStatus):
+                      XCTAssertEqual(createDump.UID, dumpStatus.UID)
+                    case .failure:
+                        XCTFail("Failed to request dump creation")
```
Comment (curquiza) [inline Tests/MeiliSearchIntegrationTests/DumpsTests.swift:51]: Weird indentation

**Review by curquiza — CHANGES_REQUESTED** [review]
Thanks for this work 🙂
Tiny changes requested

**Comment by ppamorim** [comment]
@curquiza All fixed! :)

**Comment by ppamorim** [comment]
bors try

**Review by eskombro — DISMISSED** [review]
👌 Thanks @ppamorim for this implementation!

File: Sources/MeiliSearch/Model/Dump.swift [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        public func encode(to encoder: Encoder) throws {
+            var container = encoder.singleValueContainer()
+            switch self {
+            case .processing:
+                try container.encode("processing")
+            case .dumpProcessFailed:
+                try container.encode("dump_process_failed")
+            case .done:
+                try container.encode("done")
+            }
```
Comment (bidoubiwa) [inline Sources/MeiliSearch/Model/Dump.swift:55]: What is this for? Why do we need an enum for that? 

This seems like a lot of code for something that is just a string
```
case processing
        case dumpProcessFailed
        case done
```

**Review by bidoubiwa — COMMENTED** [review]
Hey ! Thanks a lot for this PR :) I have a few questions of implementation that I do not understand.

File: Sources/MeiliSearch/Dumps.swift [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                    completion(.failure(MeiliSearch.Error.dataNotFound))
+                    return
+                }
+
+                do {
+                    let result: Dump = try Constants.customJSONDecoder.decode(Dump.self, from: data)
+                    completion(.success(result))
+                } catch {
+                    completion(.failure(error))
+                }
```
[the remark this answers is not in the document]
Comment (ppamorim) [reply Sources/MeiliSearch/Dumps.swift:69]: This parses the data that comes from the server if the decode fails an error `Swift.Error` returns in the closure.

**Review by bidoubiwa — APPROVED** [review]

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
bors try

**Review by curquiza — DISMISSED** [review]

**Comment by curquiza** [comment]
bors merge

**Comment by curquiza** [comment]
@ppamorim there is random fail with dumps in the tests -> https://github.com/meilisearch/meilisearch-swift/runs/1257819347

**Review by bidoubiwa — APPROVED** [review]

File: Tests/MeiliSearchUnitTests/DumpsTests.swift [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+        let data = json.data(using: .utf8)!
+
+        let stubDump: Dump = try! Constants.customJSONDecoder.decode(Dump.self, from: data)
+
+        session.pushData(json)
+
+        // Start the test with the mocked server
+
+        let expectation = XCTestExpectation(description: "Get settings")
```
Comment (curquiza) [inline Tests/MeiliSearchUnitTests/DumpsTests.swift:33]:
```suggestion
        let expectation = XCTestExpectation(description: "Create dump")
```

File: Tests/MeiliSearchUnitTests/DumpsTests.swift [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+        let stubDump: Dump = try! Constants.customJSONDecoder.decode(Dump.self, from: data)
+
+        session.pushData(json)
+
+        // Start the test with the mocked server
+
+        let UID: String = "20200929-114144097"
+
+        let expectation = XCTestExpectation(description: "Get settings")
```
Comment (curquiza) [inline Tests/MeiliSearchUnitTests/DumpsTests.swift:70]:
```suggestion
        let expectation = XCTestExpectation(description: "Get the dump status")
```

File: Tests/MeiliSearchUnitTests/DumpsTests.swift [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+        self.client.createDump { result in
+            switch result {
+            case .success(let dump):
+                XCTAssertEqual(stubDump, dump)
+                expectation.fulfill()
+            case .failure:
+                XCTFail("Failed to get settings")
```
Comment (curquiza) [inline Tests/MeiliSearchUnitTests/DumpsTests.swift:41]:
```suggestion
                XCTFail("Failed to create dump")
```

File: Tests/MeiliSearchUnitTests/DumpsTests.swift [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+        let expectation = XCTestExpectation(description: "Get settings")
+
+        self.client.getDumpStatus(UID: UID) { result in
+            switch result {
+            case .success(let dump):
+                XCTAssertEqual(stubDump, dump)
+                expectation.fulfill()
+            case .failure:
+                XCTFail("Failed to get settings")
```
Comment (curquiza) [inline Tests/MeiliSearchUnitTests/DumpsTests.swift:78]:
```suggestion
                XCTFail("Failed to get the dump status")
```

**Review by curquiza — CHANGES_REQUESTED** [review]
Sorry not to have seen this before... 😅

**Review by curquiza — DISMISSED** [review]

**Comment by curquiza** [comment]
bors merge

**Comment by curquiza** [comment]
tests still failing:
```
/Users/runner/work/meilisearch-swift/meilisearch-swift/Tests/MeiliSearchIntegrationTests/DumpsTests.swift:57: error: -[MeiliSearchIntegrationTests.DumpsTests testGetDumpStatus] : failed - Failed to request dump creation
/Users/runner/work/meilisearch-swift/meilisearch-swift/Tests/MeiliSearchIntegrationTests/DumpsTests.swift:61: error: -[MeiliSearchIntegrationTests.DumpsTests testGetDumpStatus] : Asynchronous wait failed: Exceeded timeout of 10 seconds, with unfulfilled expectations: "Request dump creation".
Test Case '-[MeiliSearchIntegrationTests.DumpsTests testGetDumpStatus]' failed (10.867 seconds).
Test Suite 'DumpsTests' failed at 2020-10-15 16:12:53.560.
	 Executed 2 tests, with 2 failures (0 unexpected) in 10.872 (10.872) seconds
```

**Comment by ppamorim** [comment]
@curquiza Tests are locally working, I am not sure why bors failed.

**Comment by ppamorim** [comment]
bors merge

**Comment by ppamorim** [comment]
bors try

**Comment by ppamorim** [comment]
bors try

**Comment by ppamorim** [comment]
@curquiza it was the test to create the dump running in less than 1s. I think the first test could be removed.

**Comment by ppamorim** [comment]
@curquiza I think we should remove the first test. The second test does the same.

**Comment by ppamorim** [comment]
bors try

File: Tests/MeiliSearchIntegrationTests/DumpsTests.swift [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+    override func setUp() {
+        super.setUp()
+        if client == nil {
+            client = try! MeiliSearch(Config.default(apiKey: "masterKey"))
+        }
+        pool(client)
+    }
+
+    func testCreateDumpGetDumpStatus() {
```
Comment (curquiza) [inline Tests/MeiliSearchIntegrationTests/DumpsTests.swift:20]:
```suggestion
    func testCreateAndGetDump() {
```

File: Tests/MeiliSearchIntegrationTests/DumpsTests.swift [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+                expectation.fulfill()
+            }
+        }
+
+        self.wait(for: [expectation], timeout: 10.0)
+
+    }
+
+    static var allTests = [
+        ("testCreateDumpGetDumpStatus", testCreateDumpGetDumpStatus)
```
Comment (curquiza) [inline Tests/MeiliSearchIntegrationTests/DumpsTests.swift:51]:
```suggestion
        ("testCreateAndGetDump", testCreateAndGetDump)
```

**Review by curquiza — COMMENTED** [review]
Small naming change 🙂

**Comment by ppamorim** [comment]
bors try

**Review by curquiza — APPROVED** [review]

**Comment by curquiza** [comment]
bors merge

**Comment by ppamorim** [comment]
@curquiza Thank you very much!

**Pull request merged**